camel/agents/chat_agent.py

```diff
@@ -499,220 +582,217 @@ def step(
         if isinstance(input_message, str):
             input_message = BaseMessage.make_user_message(
                 role_name='User', content=input_message
             )
 
-        if "llama" in self.model_type.lower():
-            if (
-                self.model_backend.model_config_dict.get("tools", None)
-                and not self.tool_prompt_added
-            ):
-                tool_prompt = self._generate_tool_prompt(self.tool_schema_list)
+        # Handle tool prompt injection if needed
+        if self._needs_tool_prompt() and not self.tool_prompt_added:
+            self._inject_tool_prompt()
 
-                tool_sys_msg = BaseMessage.make_assistant_message(
-                    role_name="Assistant",
-                    content=tool_prompt,
-                )
+        # Add user input to memory
+        self.update_memory(input_message, OpenAIBackendRole.USER)
 
-                self.update_memory(tool_sys_msg, OpenAIBackendRole.SYSTEM)
-                self.tool_prompt_added = True
+        # Record function calls made during the session
+        tool_call_records: List[FunctionCallingRecord] = []
 
-            self.update_memory(input_message, OpenAIBackendRole.USER)
+        # Single-step mode
+        if single_iteration:
+            return self._handle_step(response_format, tool_call_records, False)
 
-            tool_call_records: List[FunctionCallingRecord] = []
-            while True:
-                # Check if token has exceeded
-                try:
-                    openai_messages, num_tokens = self.memory.get_context()
-                except RuntimeError as e:
-                    return self._step_token_exceed(
-                        e.args[1], tool_call_records, "max_tokens_exceeded"
-                    )
+        # Multi-step mode: loop until a final response
+        if (
+            self.model_backend.model_config_dict.get("tool_choice")
+            == "required"
+        ):
+            raise ValueError(
+                "`tool_choice` cannot be set to `required` for multi-step"
+                " mode. To proceed, set `single_iteration` to `True`."
+            )
 
-                (
-                    response,
-                    output_messages,
-                    finish_reasons,
-                    usage_dict,
-                    response_id,
-                ) = self._step_model_response(openai_messages, num_tokens)
-                # If the model response is not a function call, meaning the
-                # model has generated a message response, break the loop
-                if (
-                    not self.is_tools_added()
-                    or not isinstance(response, ChatCompletion)
-                    or "</function>" not in response.choices[0].message.content  # type: ignore[operator]
-                ):
-                    break
+        return self._handle_step(response_format, tool_call_records, True)
```

Comment:
Some small suggestion to keep the code clean: Here I think we can directly return self._handle_step(response_format, tool_call_records, single_iteration), and leave the tool required check in _handle_step() method. And better keep using single_iteration argument in both two methods.